# Credit Risk Resampling Techniques

In [87]:
import warnings
warnings.filterwarnings('ignore')

In [88]:
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter
from sklearn.preprocessing import LabelEncoder, StandardScaler

# Read the CSV into DataFrame

In [89]:
# Load the data
file_path = Path('lending_data.csv')
df = pd.read_csv(file_path)
# df.head()
df.tail()

loan_size  interest_rate homeowner  borrower_income  debt_to_income  \
77531    19100.0         11.261       own            86600        0.653580   
77532    17700.0         10.662  mortgage            80900        0.629172   
77533    17600.0         10.595      rent            80300        0.626401   
77534    16300.0         10.068  mortgage            75300        0.601594   
77535    15600.0          9.742  mortgage            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt loan_status  
77531               12                 2       56600   high_risk  
77532               11                 2       50900   high_risk  
77533               11                 2       50300   high_risk  
77534               10                 2       45300   high_risk  
77535                9                 2       42300   high_risk

In [90]:
df.dtypes

loan_size           float64
interest_rate       float64
homeowner            object
borrower_income       int64
debt_to_income      float64
num_of_accounts       int64
derogatory_marks      int64
total_debt            int64
loan_status          object
dtype: object

In [91]:
label_encoder = LabelEncoder()

In [92]:
# Fitting and encoding the columns with the LabelEncoder

# loan status column
label_encoder.fit(df["loan_status"])
df["loan_status"] = label_encoder.transform(df["loan_status"])

In [93]:
df.head(20)

loan_size  interest_rate homeowner  borrower_income  debt_to_income  \
0     10700.0          7.672       own            52800        0.431818   
1      8400.0          6.692       own            43600        0.311927   
2      9000.0          6.963      rent            46100        0.349241   
3     10700.0          7.664       own            52700        0.430740   
4     10800.0          7.698  mortgage            53000        0.433962   
5     10100.0          7.438  mortgage            50600        0.407115   
6     10300.0          7.490  mortgage            51100        0.412916   
7      8800.0          6.857  mortgage            45100        0.334812   
8      9300.0          7.096       own            47400        0.367089   
9      9700.0          7.248      rent            48800        0.385246   
10     9300.0          7.085       own            47300        0.365751   
11     9200.0          7.015       own            46600        0.356223   
12     9200.0          7.043       own            46900        0.360341   
13     9400.0          7.101  mortgage            47400        0.367089   
14     8400.0          6.703  mortgage            43700        0.313501   
15     9400.0          7.136  mortgage            47700        0.371069   
16     8900.0          6.897       own            45500        0.340659   
17     8300.0          6.648  mortgage            43200        0.305556   
18     8000.0          6.539  mortgage            42100        0.287411   
19    10200.0          7.460      rent            50800        0.409449   

    num_of_accounts  derogatory_marks  total_debt  loan_status  
0                 5                 1       22800            1  
1                 3                 0       13600            1  
2                 3                 0       16100            1  
3                 5                 1       22700            1  
4                 5                 1       23000            1  
5                 4                 1       20600            1  
6                 4                 1       21100            1  
7                 3                 0       15100            1  
8                 3                 0       17400            1  
9                 4                 0       18800            1  
10                3                 0       17300            1  
11                3                 0       16600            1  
12                3                 0       16900            1  
13                3                 0       17400            1  
14                3                 0       13700            1  
15                3                 0       17700            1  
16                3                 0       15500            1  
17                2                 0       13200            1  
18                2                 0       12100            1  
19                4                 1       20800            1

In [94]:
# Binary encoding using Pandas (single column)
loans_binary_encoded = pd.get_dummies(df, columns=["homeowner"])
loans_binary_encoded.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  homeowner_mortgage  \
0                 1       22800            1                   0   
1                 0       13600            1                   0   
2                 0       16100            1                   0   
3                 1       22700            1                   0   
4                 1       23000            1                   1   

   homeowner_own  homeowner_rent  
0              1               0  
1              1               0  
2              0               1  
3              1               0  
4              0               0

In [95]:
loans_binary_encoded.dtypes

loan_size             float64
interest_rate         float64
borrower_income         int64
debt_to_income        float64
num_of_accounts         int64
derogatory_marks        int64
total_debt              int64
loan_status             int64
homeowner_mortgage      uint8
homeowner_own           uint8
homeowner_rent          uint8
dtype: object

In [96]:
# Saving the encoded dataset
file_path = Path("lending_data_encoded.csv")
loans_binary_encoded.to_csv(file_path, index=False)

In [97]:
 # Creating the scaler instance
#data_scaler = StandardScaler()

In [98]:
#data_scaler.fit(loans_binary_encoded)

In [99]:
#StandardScaler(copy=True, with_mean=True, with_std=True)

In [100]:
# Transforming the data
#loans_data_scaled = data_scaler.transform(loans_binary_encoded)
#loans_data_scaled[:5]

In [101]:
# Create our target
#y = loans_binary_encoded.loc[:,'loan_status'].copy().to_frame()
y = loans_binary_encoded['loan_status']
# Cresate our features
X = loans_binary_encoded.drop(columns="loan_status")
# Create our target
#y = df["loan_status"]

In [102]:
X.dtypes

loan_size             float64
interest_rate         float64
borrower_income         int64
debt_to_income        float64
num_of_accounts         int64
derogatory_marks        int64
total_debt              int64
homeowner_mortgage      uint8
homeowner_own           uint8
homeowner_rent          uint8
dtype: object

In [103]:
# Check the balance of our target values but y['loan_status'.value_counts() doesn't work]
#y['loan_status'].value_counts()
y.value_counts()

1    75036
0     2500
Name: loan_status, dtype: int64

In [104]:
# Create X_train, X_test, y_train, y_test
# YOUR CODE HERE
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test= train_test_split(X, 
                                                   y, 
                                                   random_state=1, 
                                                   stratify=y)
X_train.shape

(58152, 10)

## Data Pre-Processing

Scale the training and testing data using the `StandardScaler` from `sklearn`. Remember that when scaling the data, you only scale the features data (`X_train` and `X_testing`).

In [105]:
# Create the StandardScaler instance
from sklearn.preprocessing import StandardScaler
# YOUR CODE HERE
datascaler = StandardScaler()

In [106]:
# Fit the Standard Scaler with the training data
# When fitting scaling functions, only train on the training dataset
# YOUR CODE HERE
X_scaler = datascaler.fit(X_train)

In [107]:
# Scale the training and testing data
# YOUR CODE HERE
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

# Simple Logistic Regression

In [108]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver='lbfgs', random_state=1)
model.fit(X_train_scaled, y_train)

LogisticRegression(random_state=1)

In [109]:
# Calculated the balanced accuracy score
from sklearn.metrics import balanced_accuracy_score
y_pred = model.predict(X_test_scaled)
balanced_accuracy_score(y_test, y_pred)

0.9892813049736127

In [110]:
# Display the confusion matrix
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_pred)

array([[  615,    10],
       [  102, 18657]])

In [111]:
# Print the imbalanced classification report
from imblearn.metrics import classification_report_imbalanced
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.86      0.98      0.99      0.92      0.99      0.98       625
          1       1.00      0.99      0.98      1.00      0.99      0.98     18759

avg / total       0.99      0.99      0.98      0.99      0.99      0.98     19384



# Oversampling

In this section, you will compare two oversampling algorithms to determine which algorithm results in the best performance. You will oversample the data using the naive random oversampling algorithm and the SMOTE algorithm. For each algorithm, be sure to complete the folliowing steps:

1. View the count of the target classes using `Counter` from the collections library. 
3. Use the resampled data to train a logistic regression model.
3. Calculate the balanced accuracy score from sklearn.metrics.
4. Print the confusion matrix from sklearn.metrics.
5. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.

Note: Use a random state of 1 for each sampling algorithm to ensure consistency between tests

### Naive Random Oversampling

In [112]:
# view count of target
# YOUR CODE HERE
Counter(y_train)



Counter({1: 56277, 0: 1875})

In [113]:
# Resample the training data with the RandomOversampler
# YOUR CODE HERE
from imblearn.over_sampling import RandomOverSampler
# View the count of target classes with Counter
# YOUR CODE HERE
from imblearn.over_sampling import RandomOverSampler

ros = RandomOverSampler(random_state=1)
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)
Counter(y_resampled)

Counter({1: 56277, 0: 56277})

In [114]:
# Train the Logistic Regression model using the resampled data
# YOUR CODE HERE
from sklearn.linear_model import LogisticRegression

model = LogisticRegression(solver='lbfgs', random_state=1)
model.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [115]:
# Calculated the balanced accuracy score
# YOUR CODE HERE
from sklearn.metrics import balanced_accuracy_score

balanced_accuracy_score(y_test, y_pred)

0.9892813049736127

In [116]:
# Display the confusion matrix
# YOUR CODE HERE
from sklearn.metrics import confusion_matrix

y_pred = model.predict(X_test)
confusion_matrix(y_test, y_pred)

array([[  622,     3],
       [  106, 18653]])

In [117]:
# Print the imbalanced classification report
# YOUR CODE HERE
from imblearn.metrics import classification_report_imbalanced

print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.85      1.00      0.99      0.92      0.99      0.99       625
          1       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       1.00      0.99      1.00      0.99      0.99      0.99     19384



### SMOTE Oversampling

In [118]:
# Resample the training data with SMOTE
# YOUR CODE HERE

# View the count of target classes with Counter
# YOUR CODE HERE
from imblearn.over_sampling import SMOTE

X_resampled, y_resampled = SMOTE(random_state=1, sampling_strategy=1.0).fit_resample(
    X_train, y_train
)
from collections import Counter

Counter(y_resampled)

Counter({1: 56277, 0: 56277})

In [119]:
# Train the Logistic Regression model using the resampled data
# YOUR CODE HERE
model = LogisticRegression(solver='lbfgs', random_state=1)
model.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [120]:
# Calculated the balanced accuracy score
# YOUR CODE HERE
y_pred = model.predict(X_test)
balanced_accuracy_score(y_test, y_pred)

0.9948279972279972

In [121]:
# Display the confusion matrix
# YOUR CODE HERE
confusion_matrix(y_test, y_pred)

array([[  622,     3],
       [  104, 18655]])

In [122]:
# Print the imbalanced classification report
# YOUR CODE HERE
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.86      1.00      0.99      0.92      0.99      0.99       625
          1       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       1.00      0.99      1.00      0.99      0.99      0.99     19384



# Undersampling

In this section, you will test an undersampling algorithm to determine which algorithm results in the best performance compared to the oversampling algorithms above. You will undersample the data using the Cluster Centroids algorithm and complete the folliowing steps:

1. View the count of the target classes using `Counter` from the collections library. 
3. Use the resampled data to train a logistic regression model.
3. Calculate the balanced accuracy score from sklearn.metrics.
4. Display the confusion matrix from sklearn.metrics.
5. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.

Note: Use a random state of 1 for each sampling algorithm to ensure consistency between tests

In [123]:
# Resample the data using the ClusterCentroids resampler
# YOUR CODE HERE
from imblearn.under_sampling import ClusterCentroids
cc = ClusterCentroids(random_state=1)
X_resampled, y_resampled = cc.fit_resample(X_train, y_train)
# View the count of target classes with Counter
# YOUR CODE HERE

from collections import Counter
Counter(y_resampled)

Counter({0: 1875, 1: 1875})

In [124]:
# Train the Logistic Regression model using the resampled data
# YOUR CODE HERE
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver='lbfgs', random_state=1)
model.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [125]:
# Calculate the balanced accuracy score
# YOUR CODE HERE
from sklearn.metrics import balanced_accuracy_score
balanced_accuracy_score(y_test, y_pred)

0.9948279972279972

In [126]:
# Display the confusion matrix
# YOUR CODE HERE
from sklearn.metrics import confusion_matrix
y_pred = model.predict(X_test)
confusion_matrix(y_test, y_pred)

array([[  607,    18],
       [  102, 18657]])

In [127]:
# Print the imbalanced classification report
# YOUR CODE HERE
from imblearn.metrics import classification_report_imbalanced
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.86      0.97      0.99      0.91      0.98      0.96       625
          1       1.00      0.99      0.97      1.00      0.98      0.97     18759

avg / total       0.99      0.99      0.97      0.99      0.98      0.97     19384



# Combination (Over and Under) Sampling

In this section, you will test a combination over- and under-sampling algorithm to determine if the algorithm results in the best performance compared to the other sampling algorithms above. You will resample the data using the SMOTEENN algorithm and complete the folliowing steps:

1. View the count of the target classes using `Counter` from the collections library. 
3. Use the resampled data to train a logistic regression model.
3. Calculate the balanced accuracy score from sklearn.metrics.
4. Display the confusion matrix from sklearn.metrics.
5. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.

Note: Use a random state of 1 for each sampling algorithm to ensure consistency between tests

In [128]:
# Resample the training data with SMOTEENN
# YOUR CODE HERE

# View the count of target classes with Counter
# YOUR CODE HERE
from imblearn.combine import SMOTEENN
smote_enn = SMOTEENN(random_state=0)
X_resampled, y_resampled = smote_enn.fit_resample(X, y)
Counter(y_resampled)

Counter({0: 74069, 1: 74591})

In [129]:
# Train the Logistic Regression model using the resampled data
# YOUR CODE HERE
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver='lbfgs', random_state=1)
model.fit(X_resampled, y_resampled)

LogisticRegression(random_state=1)

In [130]:
# Calculate the balanced accuracy score
# YOUR CODE HERE
from sklearn.metrics import balanced_accuracy_score
balanced_accuracy_score(y_test, y_pred)

0.9828813049736127

In [131]:
# Display the confusion matrix
# YOUR CODE HERE
from sklearn.metrics import confusion_matrix
y_pred = model.predict(X_test)
confusion_matrix(y_test, y_pred)

array([[  622,     3],
       [  107, 18652]])

In [132]:
# Print the imbalanced classification report
# YOUR CODE HERE
from imblearn.metrics import classification_report_imbalanced
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

          0       0.85      1.00      0.99      0.92      0.99      0.99       625
          1       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       1.00      0.99      1.00      0.99      0.99      0.99     19384



# Final Questions

1. Which model had the best balanced accuracy score?

   The blanced accuracy scores were comparable, however, both SMOTE Oversampling and Undersampling resulted in the best scores at 0.9948279972279972.

2. Which model had the best recall score?

    Recall score for all were same at 0.99.

3. Which model had the best geometric mean score?

    Geomertic mean was the same for all at .99 except for Undersampling at .98, so it was the only one at the lowest score. 
